src/nnvm/low_precision_pass.cc

```diff
@@ -0,0 +1,265 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ *  Copyright (c) 2016 by Contributors
+ * \file low_precision_pass.cc
+ * \brief Return new graph with amp_cast and amp_multicast operators added wherever required
+ */
+
+#include <nnvm/node.h>
+#include <nnvm/graph.h>
+#include <nnvm/pass.h>
+#include <nnvm/op_attr_types.h>
+#include <mxnet/base.h>
+#include <algorithm>
+#include <functional>
+
+namespace mxnet {
+using nnvm::Symbol;
+using nnvm::Node;
+using nnvm::NodePtr;
+using nnvm::NodeEntry;
+using nnvm::Graph;
+
+// create a node for operator : op_name with name : node_name
+NodePtr CreateNode(std::string op_name, std::string node_name) {
+  NodePtr node = Node::Create();
+  node->attrs.name = node_name;
+  if (op_name == "nullptr") {
+    node->attrs.op = nullptr;
+    // ugly workaround because VariableParam is not exposed
+    node->attrs.parsed = nnvm::Symbol::CreateVariable(node->attrs.name)
+                             .outputs[0]
+                             .node->attrs.parsed;
+  } else {
+    node->attrs.op = Op::Get(op_name);
+  }
+  return node;
+}
+
+NodePtr InsertNode(std::string op_name, std::string node_name, NodePtr current,
+                   NodeEntry previous) {
+    NodePtr node = CreateNode(op_name, node_name);
+    node->inputs.emplace_back(previous);
+    current->inputs.emplace_back(NodeEntry{node, 0, 0});
+    return node;
+}
+
+// get suffix for a node entry so that it can be used for amp_cast/amp_multicast node name
+std::string GetSuffix(const nnvm::NodeEntry &e,
+                      const std::unordered_map<Node*, NodePtr> &mirror_map) {
+  static const auto &flist_outputs =
+      nnvm::Op::GetAttr<nnvm::FListOutputNames>("FListOutputNames");
+  std::string suffix = "";
+  NodePtr mirror_node = mirror_map.at(e.node.get());
+  if (mirror_node->op() != nullptr) {
+      auto list_output_names_func = flist_outputs.get(e.node->op(), nullptr);
+      if (list_output_names_func != nullptr) {
+          std::vector<std::string> names = list_output_names_func(e.node->attrs);
+          suffix = "_" + names[e.index];
+      } else {
+          suffix = "_" + std::to_string(e.index);
+      }
+  }
+  return suffix;
+}
+
+// add amp_cast node between curr_node and input
+void AddCastNode(const nnvm::NodeEntry &e, const std::string &suffix,
+                 const nnvm::NodeEntry &input, const std::string dtype,
+                 nnvm::NodeEntryMap<NodeEntry> *mirror_entry_map,
+                 NodePtr curr_node) {
+  NodePtr cast_node =
+      InsertNode("amp_cast", e.node->attrs.name + suffix + "_amp_cast_" + dtype,
+                 curr_node, input);
+  cast_node->attrs.dict["dtype"] = dtype;
+  cast_node->op()->attr_parser(&(cast_node->attrs));
+  (*mirror_entry_map)[e] = NodeEntry{cast_node, 0, e.version};
```

Comment:
microopt: you can use std::move(cast_node) since it's not used any longer. Same below.